stormevents/nhc/track.py

```diff
@@ -1196,3 +1151,83 @@ def separate_tracks(data: DataFrame) -> Dict[str, Dict[str, DataFrame]]:
+    :return: data frame of track
+    """
+
+    return pandas.concat([df for adv_trk in tracks.values() for df in adv_trk.values()])
+
+
+def correct_ofcl_based_on_carq_n_hollandb(
+    tracks: Dict[str, Dict[str, DataFrame]]
+) -> Dict[str, Dict[str, DataFrame]]:
+    """
+    Correct official forecast using consensus track along with holland-b
+    relation
+
+    :param tracks: dictionary of forecasts for each advisory (aside from best track ``BEST``, which only has one hindcast)
+    :return: dictionary of forecasts for each advisory (aside from best track ``BEST``, which only has one hindcast) with corrected OFCL
+    """
+
+    ofcl_tracks = tracks["OFCL"]
+    carq_tracks = tracks["CARQ"]
+
+    corr_ofcl_tracks = dict()
+
+    for initial_time, forecast in ofcl_tracks.items():
+        if initial_time in carq_tracks:
+            carq_forecast = carq_tracks[initial_time]
+        else:
+            carq_forecast = carq_tracks[list(carq_tracks)[0]]
+
+        relation = HollandBRelation()
+        holland_b = relation.holland_b(
+            max_sustained_wind_speed=carq_forecast["max_sustained_wind_speed"],
+            background_pressure=carq_forecast["background_pressure"],
+            central_pressure=carq_forecast["central_pressure"],
+        )
+
+        holland_b[holland_b == numpy.inf] = numpy.nan
+        holland_b = numpy.nanmean(holland_b)
+
+        # Get CARQ from forecast hour 0 and isotach 34kt (i.e. the first item)
+        carq_ref = carq_forecast.loc[carq_forecast.forecast_hours == 0].iloc[0]
+
+        columns_of_interest = forecast[
+            ["radius_of_maximum_winds", "central_pressure", "background_pressure"]
+        ]
+        columns_of_interest[columns_of_interest == 0] = pandas.NA
+        missing = columns_of_interest.isna()
+        # Order of columns is the same as columns_of_interest
+        mrd_missing = missing.iloc[:, 0]
+        mslp_missing = missing.iloc[:, 1]
+        radp_missing = missing.iloc[:, 2]
+
+        # fill OFCL maximum wind radius with the first entry from the CARQ advisory
+        forecast.loc[mrd_missing, "radius_of_maximum_winds"] = carq_ref[
+            "radius_of_maximum_winds"
+        ]
+
+        # fill OFCL background pressure with the first entry from the CARQ advisory central pressure (at sea level)
+        forecast.loc[radp_missing, "background_pressure"] = carq_ref["central_pressure"]
+
+        # fill OFCL central pressure (at sea level) with the 3rd hour entry, preserving Holland B
+        forecast.loc[mslp_missing, "central_pressure"] = relation.central_pressure(
+            max_sustained_wind_speed=forecast.loc[
+                mslp_missing, "max_sustained_wind_speed"
+            ],
+            background_pressure=forecast.loc[mslp_missing, "background_pressure"],
+            holland_b=holland_b,
+        )
+
+        corr_ofcl_tracks[initial_time] = forecast
```

Comment:
Now the correction is stored in the `corr_ofcl_tracks` which replaces the values in the original forecast track. #66